src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/CompileExpressionsTests.cs

```diff
@@ -464,5 +466,240 @@ .maxstack  9
 }");
                 });
         }
+
+        [WorkItem(482753, "https://devdiv.visualstudio.com/DevDiv/_workitems?id=482753")]
+        [Fact]
+        public void LocalsInAsync()
+        {
+            var source =
+@"using System;
+using System.Threading.Tasks;
+class C
+{
+    static Task<object> E(object o, Func<object, bool> p)
+    {
+        throw new NotImplementedException();
+    }
+    object F()
+    {
+        throw new NotImplementedException();
+    }
+    Task G(object o)
+    {
+        throw new NotImplementedException();
+    }
+    async Task M(object x)
+    {
+        var z = await E(F(), y => x == y);
+#line 999
+        await G(z);
+    }
+}";
+            // Test with CompileExpression rather than CompileExpressions
+            // so field references in IL are named.
+            // Debug build.
+            var comp = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugDll, references: new[] { SystemCoreRef });
+            WithRuntimeInstance(
+                comp,
+                references: null,
+                includeLocalSignatures: true,
+                includeIntrinsicAssembly: false,
+                validator: runtime =>
+                {
+                    var context = CreateMethodContext(runtime, "C.<M>d__3.MoveNext()", atLineNumber: 999);
+                    string error;
+                    var testData = new CompilationTestData();
+                    var result = context.CompileExpression("z ?? x", out error, testData);
+                    Assert.NotNull(result.Assembly);
+                    Assert.Null(error);
+                    testData.GetMethodData("<>x.<>m0").VerifyIL(
+@"{
+  // Code size       22 (0x16)
+  .maxstack  2
+  .locals init (int V_0,
+                System.Runtime.CompilerServices.TaskAwaiter<object> V_1,
+                C.<M>d__3 V_2,
+                System.Runtime.CompilerServices.TaskAwaiter V_3,
+                System.Exception V_4)
+  IL_0000:  ldarg.0
+  IL_0001:  ldfld      ""object C.<M>d__3.<z>5__2""
+  IL_0006:  dup
+  IL_0007:  brtrue.s   IL_0015
+  IL_0009:  pop
+  IL_000a:  ldarg.0
+  IL_000b:  ldfld      ""C.<>c__DisplayClass3_0 C.<M>d__3.<>8__1""
+  IL_0010:  ldfld      ""object C.<>c__DisplayClass3_0.x""
+  IL_0015:  ret
+}");
+                });
+            // Release build.
+            comp = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseDll, references: new[] { SystemCoreRef });
+            {
+                // Note from MoveNext() below that local CS$<>8__locals0 should not be
+                // used in the compiled expression to access the display class since that
+                // local is only set the first time through MoveNext() (see loc.2 below).
```

Comment:
>local is only set the first time through MoveNext() [](start = 19, length = 51)

Do I understand this correctly that the synthesized local is NOT initialized until certain point in MoveNext so to access the parameter we want to read it directly and thru the display class? If so, I'd change the comment to say "... to access the parameter thru the display class since the we need to read the parameter value before the display class is initialized"

Reply:
The display class local `loc.2` is only set the first time through `MoveNext`, so in this case we should access the local `x` from the field on `this` (since `this` is a display class) rather than field on the display class local.